activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -8,12 +8,13 @@ def initialize
 
       # Converts an arel AST to SQL
       def to_sql(arel, binds = [])
-        if arel.respond_to?(:ast)
+        sql = if arel.respond_to?(:ast)
           collected = visitor.accept(arel.ast, collector)
           collected.compile(binds, self)
         else
           arel
```

Comment:
I'm not sure how we get here, off the top of my head, but if we do, this is again not our string to freeze.

It does seem reasonable to do a `dup.freeze` here though. (I'd move the `freeze` into each branch of the conditional, to keep it right next to the `dup`.)